KEP 2232: add Suspend Job in implementable status (#2234)
kubernetes/enhancements: keps/sig-apps/2232-jobs-stopped/README.md

its execution. Currently, this is impossible; as a workaround, users can delete
and re-create the Job, but this will also delete Job metadata such as the list
of successful/failed Pods and logs. Therefore, it is desirable to stop Jobs and
resume them later when convenient.

### Goals

- Allow stopping and resuming Jobs.
- Allow indefinitely delaying the creation of Pods owned by a Job.

### Non-Goals

- Being able to stop and resume jobs makes Job preemption, all-or-nothing
  scheduling, and Job queueing possible. However, we don't intend to create
  such higher-level controllers as a part of this KEP.
- It might be useful to restrict who can set/modify the `stopped` flag.
  However, we don't intend to create the validating and/or mutating webhooks
  necessary to achieve that as a part of this KEP.

## Proposal

### User Stories

#### Story 1

Consider a cloud provider where servers are cheaper in the night. My Job takes
several days to complete, so I'd like to stop my Job early in the morning every
day and resume it after dark to save money. However, I don't want to delete and
re-create my Job every day as I don't want to lose track of completed Pods,
logs, and other metadata.

> #### Story 2

Reviewer: I see you still have the mutating webhook story, I don't think this is convincing, especially the part where "a cluster that can only run one Job at any given time"

I'm the system administrator of a cluster than can only run one Job at any
given time. I'd like my users to submit their Jobs whenever they want, but I
want control over whose Job gets to run on the cluster. With a custom mutating
webhook that adds `stopped: true` to all user-submitted Jobs, I can
indefinitely delay Pod creation.

### Notes/Constraints/Caveats

System administrators may want to restrict who can set/modify the `stopped`
field when creating or updating Jobs. This can be achieved with validating
and/or mutating webhooks.

> When a Job is stopped, the time will be recorded as a part of the Job status.

Reviewer: ```suggestion
When a Job is stopped, the time will be recorded as part of the Job status.
```

	// false to true), the Job controller will delete all Pods associated with
	// this Job. This is an alpha field and requires enabling the StoppedJob
	// feature gate. Defaults to false.
	Stopped *bool `json:"stopped,omitempty"`

	...
}
```

As described in the comment, when the boolean is set to true, the
controller-manager will abstain from creating Pods even if the Job's
parallelism requirements aren't met. If the Job is already active and is
updated with `stopped: true`, all Pods owned by this Job will be deleted.

Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
"Complete" and "Failed", we propose adding a new condition called "Stopped"
that will be a part of the Job's status:

```go
type JobStatus struct {
	// The latest available observations of an object's current state. Can be
	// one of "Stopped", "Failed", or "Complete".
	Conditions []JobCondition `json:"conditions,omitempty"`

	...
}

type JobConditionType string

// These are valid conditions of a job.
const (
	// JobStopped means the job has been stopped.

> 	JobStopped JobConditionType = "Stopped"

Reviewer: A couple of questions:
1) should the job transition to another new condition when the job is unstopped? 
2) will this condition exist in case the job was created in stopped state from the get go?

Author: >should the job transition to another new condition when the job is unstopped?

I don't think so. There is currently no status when a job is active. I think a resumed job should be as similar to a never-stopped job as possible.

Besides, the condition should be statement about the present, not the past ("the job *was* unstopped, the job *is* active/stopped/complete").

>will this condition exist in case the job was created in stopped state from the get go?

Yes.

Reviewer: the other two conditions are terminal, as in once the job reaches them it will never transition out. The stopped one is different in that you can transition out of it. So if we are not transitioning to another condition after the job has been resumed, then a user can't rely on the job condition to check whether or not the job is stopped, it should always check the flag. Which brings two questions: what is the value of having a stopped JobCondition?

Author: >a user can't rely on the job condition to check whether or not the job is stopped

The presence/lack of the Stopped condition will tell them if the job is stopped or not.

>what is the value of having a stopped JobCondition

There is a non-zero amount of time between updating the Job spec and the Job controller observing the change. If the native k8s Job controller is buggy or down, a Job could have `stopped: true` without a Stopped condition indefinitely.

The controller will add the Stopped condition only *after* it has requested the API server to delete all pods.

---

So JobConditions also have a boolean Status field. We could transition from `Stopped: true` to `Stopped: false` in the status when unstopped. Jobs that have never been stopped will not have the condition at all. Updated the KEP to reflect this.

Reviewer: ok, so we are transitioning to another state: condition.Status=false when unstopping the job, this clarified it. thanks.

	// JobComplete means the job has completed its execution.
	JobComplete JobConditionType = "Complete"
	// JobFailed means the job has failed its execution.
	JobFailed JobConditionType = "Failed"
)
```

When a Job is stopped or created in the stopped state, a "Stopped" event will
be recorded. Similarly, when a Job is resumed from its stopped state, a
"Resumed" event will be recorded.

> In addition to the new JobConditionType, a new `StopTime` field is also

Reviewer: I don't think we need this. The JobCondition has a timestamp: https://github.com/kubernetes/kubernetes/blob/a04b6e4b1671810ede5b8cacf4527741781d6fb9/staging/src/k8s.io/api/batch/v1/types.go#L188

Author: good catch, that should work! Do you think this will be obvious enough to users? I certainly didn't notice that that field can be used to infer stop time.

Reviewer: Proper documentation should be enough. We don't want to introduce a new field just because a synonym one is not too obvious.

Upgrading from 1.20 and below will not change the behaviour of how Jobs work.

To make use of this feature, the `StoppedJob` feature gate must be explicitly
enabled on the API server and the controller manager and the `stopped` field
must be explicitly set in the Job spec.

### Version Skew Strategy

The change is entirely limited to the control plane. Version skew across
control plane / kubelet does not change anything.

In HA clusters, version skew across different replicas in the control plane
should also work seamlessly because only one controller manager will be active
at any given time.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

* **How can this feature be enabled / disabled in a live cluster?**
  - [x] Feature gate (also fill in values in `kep.yaml`)
    - Feature gate name: `StoppedJob`
    - Components depending on the feature gate:
      - kube-apiserver
      - kube-controller-manager

* **Does enabling the feature change any default behavior?** No. Unless the
  `stopped` field is explicitly set, default behaviour will remain the same.

* **Can the feature be disabled once it has been enabled (i.e. can we roll back

>   the enablement)?** Yes. Turning off the feature gate will disable the

Reviewer 2: What will happen with jobs that has the Stopped field set?

Author: The job controller will ignore the stopped field if the feature gate isn't enabled. So if you restart the controller manager and there were stopped jobs, they will be assumed to be active (even though the stopped field is set in etcd; it will just be ignored).

I'll clarify this in the KEP.

- [Design Details](#design-details)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Scalability](#scalability)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed](#infrastructure-needed)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [x] (R) KEP approvers have approved the KEP status as `implementable`
- [x] (R) Design details are appropriately documented
- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input

> - [x] (R) Graduation criteria is in place

Author: should this be checked?

Reviewer: yes, I added one.

System administrators may want to restrict who can set/modify the `stopped`
field when creating or updating Jobs. This can be achieved with validating
and/or mutating webhooks.

When a Job is stopped, the time is recorded as part of the Job status.  Users
can infer how long a Job has been paused for through this field. This can be
useful when making decisions around which Job should be resumed.

A Job that is complete cannot be stopped.

### Risks and Mitigations

1. Stopping an active Job deletes *all* active pods. Users must design their
   application to gracefully handle this.

## Design Details

We propose adding a `stopped` field to the JobSpec API:

```go
type JobSpec struct {
	// Stopped specifies whether the Job controller should create Pods or not. If
	// a Job is created in the stopped state, no Pods are created by the Job
	// controller. If a Job enters the stopped state (i.e. the flag goes from
	// false to true), the Job controller will delete all active Pods associated
	// with this Job. Users must design their workload to gracefully handle this.
	// This is an alpha field and requires enabling the StoppedJob feature gate.
	// Defaults to false.

> 	Stopped *bool `json:"stopped,omitempty"`

Reviewer 3: Why pointer bool? I don't think it matters to have 3 states expressed, we only care about yes/no.

- Deprecate the flag

**For non-optional features moving to GA, the graduation criteria must include 
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
-->

### Upgrade / Downgrade Strategy

Upgrading from 1.20 and below will not change the behaviour of how Jobs work.

To make use of this feature, the `StoppedJob` feature gate must be explicitly
enabled on the API server and the controller manager and the `stopped` field
must be explicitly set in the Job spec.

### Version Skew Strategy

The change is entirely limited to the control plane. Version skew across
control plane / kubelet does not change anything.

In HA clusters, version skew across different replicas in the control plane
should also work seamlessly because only one controller manager will be active
at any given time.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

* **How can this feature be enabled / disabled in a live cluster?**
  - [x] Feature gate (also fill in values in `kep.yaml`)
    - Feature gate name: `StoppedJob`
    - Components depending on the feature gate:

>       - kube-apiserver

Reviewer 3: Why kube-apiserver it only applies to kcm?

    - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
    - [Beta -&gt; GA Graduation](#beta---ga-graduation)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Scalability](#scalability)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed](#infrastructure-needed)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [x] (R) KEP approvers have approved the KEP status as `implementable`
- [x] (R) Design details are appropriately documented
- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [x] (R) Graduation criteria is in place
- [x] (R) Production readiness review completed

> - [ ] Production readiness review approved

Reviewer 2: nit: you can probably mark this (it will be approved once this PR merges)

Reviewer: done

of this feature?**
  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
  implementation difficulties, etc.).

### Dependencies

_This section must be completed when targeting beta graduation to a release._

* **Does this feature depend on any specific services running in the cluster?**
  Think about both cluster-level services (e.g. metrics-server) as well
  as node-level agents (e.g. specific version of CRI). Focus on external or
  optional services that are needed. For example, if this feature depends on
  a cloud provider API, or upon an external software-defined storage or network
  control plane.

  For each of these, fill in the following—thinking about running existing user workloads
  and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:

-->

### Scalability

* **Will enabling / using this feature result in any new API calls?** Yes. When
  a Job is stopped, a `PATCH jobs` request will be necessary to update the

>   field when it is time to resume the Job.

Reviewer 2: That is probably less of an issue then the fact that all pods will be deleted then (and similarly created when unstopping).

Can you please add that?

Reviewer: done.

in the right order at the right time.

I can write a higher-level Job queueing controller to do this based on external
factors. For example, the controller could choose to simply unpause Jobs in the
FIFO order. Alternatively, Jobs could be assigned priority and, just like
kube-scheduler, the controller can make a decision based on the stopped Job
queue (including Job preemption). Each Job could request a different amount of
resources, so the higher-level controller may also want to resize the cluster
to just fit the Job it's going to run. Regardless of what logic the controller
uses to queue jobs, being able to stop Jobs indefinitely and then resume them
later is important.

### Notes/Constraints/Caveats

System administrators may want to restrict who can set/modify the `stopped`
field when creating or updating Jobs. This can be achieved with validating
and/or mutating webhooks.

When a Job is stopped, the time will be recorded as part of the Job status.
Users can infer how long a Job has been paused for through this field. This can
be useful when making decisions around which Job should be resumed.

A Job that is complete cannot be stopped.

> Stopping Jobs will be gated behind a feature gate and will need to be turned on

Reviewer 4: This is standard for any alpha feature, so no need to mention here.

explicitly.

### Risks and Mitigations

1. Stopping an active Job will delete *all* active pods. Users must design
   their application to gracefully handle this.

## Design Details

We propose adding a `stopped` field to the JobSpec API:

```go
type JobSpec struct {
	// Stopped specifies whether the Job controller should create Pods or not. If
	// a Job is created in the stopped state, no Pods are created by the Job
	// controller. If a Job enters the stopped state (i.e. the flag goes from
	// false to true), the Job controller will delete all Pods associated with
	// this Job. This is an alpha field and requires enabling the StoppedJob
	// feature gate. Defaults to false.
	Stopped *bool `json:"stopped,omitempty"`

	...
}
```

As described in the comment, when the boolean is set to true, the
controller-manager will abstain from creating Pods even if the Job's
parallelism requirements aren't met. If the Job is already active and is

> updated with `stopped: true`, all Pods owned by this Job will be sent a SIGTERM

Reviewer 4: prefer to use active voice:

the job controller calls Delete on all of its active Pods. This causes kubelet to send a SIGTERM signal and completely remove the Pod after its graceful termination period is honored. The controller does not count terminated Pods towards completions.

You could mention that this behavior is similar to reducing parallelism.

	Type JobConditionType `json:"type"`

	// Status of the condition, one of True, False, Unknown.
	Status v1.ConditionStatus `json:"status"`

	// Last time the condition transit from one status to another.
	// +optional
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`

	...
}

type JobConditionType string

// These are valid conditions of a job.
const (
	// JobStopped means the job has been stopped.
	JobStopped JobConditionType = "Stopped"
	// JobComplete means the job has completed its execution.
	JobComplete JobConditionType = "Complete"
	// JobFailed means the job has failed its execution.
	JobFailed JobConditionType = "Failed"
)
```

To determine if a Job has been stopped, users must look for the JobCondition
with `Type` as "Stopped" and `Status` as true. If such a JobCondition does not
exist, the Job has never been stopped. Otherwise, if the `Status` is false, the
Job was previously stopped and then resumed. Inferring this from the Job spec
is not recommended as the Job controller may not have seen the update yet. When
a Job is stopped, the Job controller will set the condition only after all Pods
are terminating or have been deleted.

> The Stopped JobCondition also has a `LastTransitionTime` field. This can be

Reviewer 4: Does the semantics of `.status.startTime` change?

Author: Yep, done.

used to infer how long a Job has been stopped for (if the `Status` value is
true).

When a Job is stopped or created in the stopped state, a "Stopped" event will
be recorded. Similarly, when a Job is resumed from its stopped state, a

> "Resumed" event will be recorded.

Reviewer 4: Do we get a Resumed event if the Job is created with stopped=false?

Author: No. An event is sent only when the Job transitions between the two states.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The Jobs API allows users to create Pods using the jobs controller with
specific requirements such as Pod-level parallelism, completing a certain
number of executions, and restart policies.

> This KEP proposes an enhancement that will allow stopping and resuming Jobs.

Reviewer 3: This looks mostly good, my only suggestion would be not to use `stopped` which might suggest we are halting the job, but maybe rather suspend or paused? I'm personally leaning towards the former, since that's what CronJob has so having it consistent for both batch resources would be desirable.

Author: I didn't know CronJob had a suspend field. In that case, I prefer suspend too.

@ahg-g @alculquicondor opinions before I make the change?

Reviewer: +1 to suspended. pause exist in deployment, but has completely different semantics, so -1 to that.

Reviewer 4: +1 We discussed and agreed to this at the sig meeting.

Reviewer 2: Hmm - I'm actually wondering about that.
In CronJob - we are suspending "subsequent executions":
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/batch/v1beta1/types.go#L119

Here - we are also stopping the currently running ones.
So isn't stopped better reflecting that?
@soltysh @ahg-g - thoughts?

Reviewer: I am agnostic, in https://github.com/kubernetes/kubernetes/issues/97170 this I used "suspended/stopped" :)

But I agree that stopped would then be better if the CrongJob semantics doesn't include stopping a currently running job.

Reviewer: kept it Stopped for now, pls let me know if this needs to change.

Reviewer 2: Can you add something like the lines of

We will discuss with API approver which of "stopped" or "suspended" is a better name, given existence of "suspended" in CronJob (though with slightly different semantics).

Does that make sense?

Reviewer: @lavalamp can you please give your opinion on the flag name?

Reviewer 3: I'll be in favor of suspended, but like @wojtek-t mentioned this can be resolved at api review time and updated here.

Reviewer 5: If you weren't killing the existing pods, I'd vote for "paused" or "suspended".

If you restart a job after it's been stopped, does the whole thing restart or does it count the pods that already did succeed?

If it starts over, then stop/start seems right. If it latches progress, then maybe pause/continue or suspend/resume is better.

Reviewer: it will will latch on progress made, finished pods will count towards the completion status of the job.

Reviewer 5: I think "suspend" is a good word, it strongly implies that a "resume" is possible, and also (at least to me) implies more action wrt the currently running pods than "pause" would.

"stop" is probably my least favorite, since stopped things can't always be restarted and also might start from the beginning.

(I think this is a little subjective)

I think it needs to be very clear about what happens to currently running pods, I personally would have guessed that users don't want to lose any progress those pods have made. Are you sure users won't want that?

Reviewer: > I think it needs to be very clear about what happens to currently running pods, I personally would have guessed that users don't want to lose any progress those pods have made. Are you sure users won't want that?

This is the same semantics as setting `parallelism=0`. Deleting running pods allows a job-level scheduler to "preempt" the job.

Reviewer 5: OK, that makes sense to me, I think "suspend" is a good term, and I imagine the documentation for such a scheduler will explain the potential consequences. sgtm.

Reviewer 4: The Pods still get a termination signal, so they have a chance to save their work.

Suspend implies some middle ground. I think the relative ambiguity of it will at least trigger users to read the field documentation.

Reviewer 2: @lavalamp The only drawback of suspend I see is the inconsistency with the "CronJob" semantics (where we don't kill the already running job). But if you think that's ok, I'm fine with it.

Reviewer 5: @wojtek-t Yeah, I hear that. I'm somewhat torn between a) just living with it, b) adding subtypes to "suspend" semantics (e.g., "hard" vs "soft" suspension), or c) coming up with a different word--unfortunately in a vacuum I think "suspend" is the best word :/

I lean towards option a) and fixing it with documentation. Maybe we should get a 3rd opinion though.

Reviewer 2: I don't have a better suggestion too (and I think (b) will introduce even more confusion). So I'm fine with (a) too - just wanted to ensure we all are aware of that (and we are) - which makes me happy.

Reviewer 6: (b) isn't a bad idea. It's also possible to extend this field to have soft & hard suspension (hard as default when `suspend` is set) after alpha, so let's go with (a) for now until we gather more user feedback.

uses to queue jobs, being able to stop Jobs indefinitely and then resume them
later is important.

### Notes/Constraints/Caveats

System administrators may want to restrict who can set/modify the `stopped`
field when creating or updating Jobs. This can be achieved with validating
and/or mutating webhooks.

When a Job is stopped, the time is recorded as part of the Job status.  Users
can infer how long a Job has been paused for through this field. This can be
useful when making decisions around which Job should be resumed.

A Job that is complete cannot be stopped.

### Risks and Mitigations

1. Stopping an active Job deletes *all* active pods. Users must design their
   application to gracefully handle this.

## Design Details

We propose adding a `stopped` field to the JobSpec API:

```go
type JobSpec struct {
	// Stopped specifies whether the Job controller should create Pods or not. If
	// a Job is created in the stopped state, no Pods are created by the Job
	// controller. If a Job enters the stopped state (i.e. the flag goes from
	// false to true), the Job controller will delete all Pods associated with

> 	// this Job. This is an alpha field and requires enabling the StoppedJob

Reviewer 3: controller will delete all **running** Pods, not **all** Pods, which is what you said elsewhere.

Reviewer 3: I'd also mention here explicitly that it is user's responsibility to design an application such that it can be suspended half way through, which means some of the pods will be terminated.

  - [Version Skew Strategy](#version-skew-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Scalability](#scalability)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed](#infrastructure-needed)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Reviewer 3: Check the necessary items.

Reviewer: done.

	...
}
```

As described in the comment, when the boolean is set to true, the
controller-manager abstains from creating Pods even if the Job's parallelism
requirements aren't met. If the Job is already active and is updated with
`stopped: true`, the Job controller calls Delete on all its active Pods. This
causes the kubelet to send a SIGTERM signal and completely remove the Pod after
its graceful termination period is honoured. Pods terminated this way are
considered a failure and the controller does not count terminated Pods towards
completions. This behaviour is similar to decreasing parallelism to zero.

Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
"Complete" and "Failed", we propose adding a new condition type called
"Stopped" as a part of the Job's status:

```go
type JobStatus struct {
	// The latest available observations of an object's current state. Can be
	// one of "Stopped", "Failed", or "Complete".
	Conditions []JobCondition `json:"conditions,omitempty"`

	...
}

// JobCondition describes current state of a job.

> type JobCondition struct {

Reviewer 2: A generic condition should be used instead:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1353

Reviewer: This struct already exists, we are not adding it.

	// Last time the condition transit from one status to another.
	// +optional
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`

	...
}

type JobConditionType string

// These are valid conditions of a job.
const (
	// JobStopped means the job has been stopped.
	JobStopped JobConditionType = "Stopped"
	// JobComplete means the job has completed its execution.
	JobComplete JobConditionType = "Complete"
	// JobFailed means the job has failed its execution.
	JobFailed JobConditionType = "Failed"
)
```

To determine if a Job has been stopped, users must look for the JobCondition
with `Type` as "Stopped" and `Status` as true. If such a JobCondition does not
exist, the Job has never been stopped. Otherwise, if the `Status` is false, the
Job was previously stopped and then resumed. Inferring this from the Job spec
is not recommended as the Job controller may not have seen the update yet. When
a Job is stopped, the Job controller sets the condition only after all Pods are
terminating or have been deleted.

The Stopped JobCondition also has a `LastTransitionTime` field. This can be
used to infer how long a Job has been stopped for (if the `Status` value is

> true).

Reviewer: out of curiosity, when updating an existing job condition from true to false, does this happen in place or as an addition entry in the Conditions list?

Reviewer 3: Since job is a finite resource, we only add new conditions. Currently, only https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/pkg/controller/job/job_controller.go#L534 and https://github.com/kubernetes/kubernetes/blob/9f2f1b8c460b4b5ec8bfcd34490fb259ebad65b0/pkg/controller/job/job_controller.go#L568

Reviewer 2: Wait - are you saying that a Job may have multiple conditions with the same type? That sounds wrong...

Reviewer: I am guessing currently no since the two conditions we now have (Failed and Completed) are terminal once the job reaches to either.

This KEP proposes adding a new condition that we can exit from (Stopped). I am not sure what is the common pattern when dealing with such conditions, do we append a new entry or update the existing one? I am guessing the former.

Reviewer 2: We update the existing one.

Reviewer 2: In other words: for a given condition type, there should always be at most one entry.

Reviewer: sg, I updated the text to emphasize that.

If a Job is created with `stopped: true`, the `StartTime` field of the Job
be set when it is resumed for the first time.

When a Job is stopped or created in the stopped state, a "Stopped" event is
recorded. Similarly, when a Job is resumed from its stopped state, a "Resumed"
event is recorded.

### Test Plan

Unit, integration, and end-to-end tests will be added to test different
scenarios including:

- Creating a Job with `stopped: true`
- Stopping an active Job and resuming it again
- Verifying that partially completed jobs remember completions after a
  stop-resume cycle

### Graduation Criteria

> N/A. Feature will be alpha in the targeted release (1.21).

Reviewer 2: Actually, specifying graduation criteria for further steps is desired at the beginning. (we can update that later, but that's the more conscious decision).

Reviewer: updated.

Author: sorry, I'll let @wojtek-t resolve these

[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
-->

### Upgrade / Downgrade Strategy

Upgrading from 1.20 and below will not change the behaviour of how Jobs work.

To make use of this feature, the `StoppedJob` feature gate must be explicitly
enabled on the API server and the controller manager and the `stopped` field
must be explicitly set in the Job spec.

### Version Skew Strategy

The change is entirely limited to the control plane. Version skew across
control plane / kubelet does not change anything.

In HA clusters, version skew across different replicas in the control plane
should also work seamlessly because only one controller manager will be active
at any given time.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

* **How can this feature be enabled / disabled in a live cluster?**
  - [x] Feature gate (also fill in values in `kep.yaml`)
    - Feature gate name: `StoppedJob`
    - Components depending on the feature gate:
      - kube-apiserver
      - kube-controller-manager

* **Does enabling the feature change any default behavior?** No. Unless the

>   `stopped` field is explicitly set, default behaviour will remain the same.

Reviewer 2: Explicitly setting "stopped" field is not a default behavior.

Please remove that part of the sentence - change to something like:

"No, using feature requires explicitly opting in by setting `stopped` field."

Reviewer: updated.

* **Can the feature be disabled once it has been enabled (i.e. can we roll back
  the enablement)?** Yes. Turning off the feature gate will disable the
  feature. Once the feature gate is turned off, the default Job controller will
  ignore the `stopped` field in all jobs, so this will cause existing stopped
  jobs to be resumed indirectly when the controller manager is restarted.

* **What happens if we reenable the feature if it was previously rolled back?**

>   All components will continue to work as expected.

Reviewer 2: What does it mean?
In particular - it's useful to explain that jobs with `stopped` field set, will get stopped by deleting all their pods.

Reviewer: updated.

  as node-level agents (e.g. specific version of CRI). Focus on external or
  optional services that are needed. For example, if this feature depends on
  a cloud provider API, or upon an external software-defined storage or network
  control plane.

  For each of these, fill in the following—thinking about running existing user workloads
  and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:

-->

### Scalability

* **Will enabling / using this feature result in any new API calls?** Yes. When
  a Job is stopped, a `PATCH jobs` request will be necessary to update the
  field when it is time to resume the Job.

* **Will enabling / using this feature result in introducing new API types?**
  No.

* **Will enabling / using this feature result in any new calls to the cloud 
  provider?** No.

* **Will enabling / using this feature result in increasing size or count of 
  the existing API objects?** Each JobSpec object will increase in size by four

>   bytes. Each JobStatus will have an additional time.Time object and a string

Reviewer 2: It's impossible to say whether it's exactly 4 or not, because it depends on the serialization. Let's just mention that a new bool fields is added.

Reviewer: updated

>   in its Conditions field when the Job is stopped.

Reviewer 2: You should use generic condition type - please update here and elsewhere accordindly.

Reviewer: as mentioned above, this is an existing JobCondition struct.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

tl;dr: make it possible to create Jobs without immediately creating Pods.

The Jobs API allows users to create Pods using the jobs controller with
specific requirements such as Pod-level parallelism, completing a certain
number of executions, and restart policies. This KEP proposes an enhancement

> that will allow higher-level orchestrators to create Jobs with the option to

Reviewer: s/orchestrators/controller just so we use k8s terminology

Reviewer: and all other places in the KEP

NOT immediately create Pods until a bit is flipped. Such a Job is called a
*Stopped Job* and is facilitated through the `stopped` field in the Job spec.

## Motivation

The Job controller tracks and manages all Jobs within Kubernetes. When a new

> Job is created by an user or a higher-level orchestrator and is seen by the

Reviewer: ```suggestion
Job is created and is seen by the
```

controller, it will immediately begin creating pods to satisfy the requirements
of the Job.

This is sometimes undesirable if a higher-level orchestrator wants to create a
Job in a "stopped" state. For instance, if a Job is created and there aren't
enough resources to schedule all Pods to nodes, the Job will be in a state
where its Pod-level parallelism requirements are not met. If the cluster is
able to scale up its resources, the Job will eventually meet such requirements,

> but this isn't useful in cases such as all-or-nothing scheduling where Pods

Reviewer 4: other than all-or-nothing, some quota-based queuing would also be possible with the stopped flag.

cannot make progress until all Pods are available. In such scenarios, in order
to optimally make use of available resources, it is preferable to wait for the

> entire Job to become schedulable before creating any Pods.

Reviewer 4: Another point is that having unschedulable Pods in the systems adds stress to the scheduler.

A consequence of this is that even high priority system Pods (that are not from Jobs) can be starved. Of course, this is only a problem if multiple Jobs are created simultaneously and continuously.

Reviewer: another one is resource quota, physical resources may be available but no quota is available for the job's namespace

requires scaling up the cluster to accommodate it, but this isn't
instantaneous. If the kube-controller-manager created all the Pods of Job C,
the kube-scheduler, which operates at the Pod-level, could schedule them to
nodes before the Pods of A or B. In Open MPI, this oversubscribes the worker
nodes, resulting in degraded performance. It might have been more efficient to
schedule the Pods of A and B first and then wait for the cluster to scale up to
18 nodes before scheduling the Pods of C.

If the Job specs had a `stopped` field set to true at the time of creation of
the object, the controller-manager could use this information and hold off from
creating the Pods of this Job. This field could be automatically set to true at
creation with a mutating webhook and then be modified by an external controller
once cluster resources become available:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: openmpi
spec:
  stopped: true  # added with a mutating webhook
  parallelism: 4
  template:
    spec:
      containers:
      - name: foo
        image: bar
        command: ["baz"]
```

[Open MPI]: https://www.open-mpi.org/

#### Story 2

> <!-- An example to motivate stopping active Jobs? -->

Reviewer: preemption perhaps

Exactly how and when a Job is resumed is left to the orchestrator (e.g. cluster
scaling succeeded, quotas are relaxed, or the Job is at the head of the queue).
Note that Pods created by the Job controller are still scheduled to nodes by
the scheduler.

### Risks and Mitigations

1. Stopping an active Job will delete *all* active pods.

2. It is important to prevent users from interfering with the proposed field
   manually in jobs managed by a higher-level orchestrator.

## Design Details

We propose adding a `stopped` field to the JobSpec API:

```go
type JobSpec struct {
	// Stopped specifies whether the job controller should create Pods or not. If
	// a Job is created in the stopped state, no Pods are created by the job
	// controller. If a Job enters the stopped state (i.e. the flag goes from
	// false to true), the job controller will delete all Pods associated with
	// this job. This field is generally used by higher-level orchestrators that
	// need more control over when the Pods of a Job are created.
	Stopped *bool

	...
}
```

As described in the comment, when the boolean is set to true, the
controller-manager will abstain from creating Pods even if the Job's

> parallelism requirements aren't met. If the Job is already active and is

Reviewer 4: Add more detail on how stop is handled when there are active Pods.

Things to think about:
- will Pods have forceful termination?
- If the Pods receive a termination signal, and they exit with a 0 exit code, wouldn't kubelet mark them as succeeded? That would cause the Job to increase it's completions counter or finish (if `.spec.completions` is not set). Maybe some of this scenarios are desirable in the case of preemption, and users are expected to create a new Job.

Reviewer: @adtac can you describe the parallelism=0 semantics? we may be ok if we simply fallback to that behavior.

Reviewer 4: It just doesn't create new Pods. The existing Pods continue running.

Author: No, decreasing parallelism will delete pods:                                                    

https://github.com/kubernetes/kubernetes/blob/faaae714274c2fd0eb6ce71a77d54106b7689e3d/pkg/     controller/job/job_controller.go#L702

which deletes the pod:

https://github.com/kubernetes/kubernetes/blob/faaae714274c2fd0eb6ce71a77d54106b7689e3d/pkg/     controller/controller_utils.go#L600

I'll post an update with the exact exit code, but I don't think it's 0.

Author: I just verified that the job's pods that get terminated do not count towards the completions. In fact, I can't even query the pod with `kubectl get`, so it looks like all references to it are removed.

Reviewer 4: I guess this makes sense. It's an explicit deletion, even if it takes some time for the actual deletion to happen.

Reviewer 4: Add the detail of how Pod termination is handled to the KEP.

However, I still have one doubt. I assume the process goes like follows:
1. Controller calls DELETE on the API
2. kube-apiserver adds the deletionTimestamp
3. Kubelet sends the termination signal
4. Containers terminate 
5. kubelet marks the pod as finished or removes the Pod?

If kubelet first marks the Pod as finished (Failed or Succeeded) and then deletes the Pod, there could be a time where the controller counts the Pod towards the Job status.
OTOH, if kubelet directly removes the Pod, then there is no chance for the Job controller to use this Pod towards status.

Reviewer: can you resolve this please.

Author: >Add the detail of how Pod termination is handled to the KEP

Done.

>kubelet marks the pod as finished or removes the Pod?

For others: we discussed this offline and concluded that the Kubelet will never mark these pods as finished:

https://github.com/kubernetes/kubernetes/blob/9da11e294fae8e26330d5c0ed5dff61cda178d77/staging/src/k8s.io/api/core/v1/types.go#L2509

Pods terminated by the system will always be marked as failed, so they will not count towards completions.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

tl;dr: make it possible to create Jobs without immediately creating Pods.

The Jobs API allows users to create Pods using the jobs controller with
specific requirements such as Pod-level parallelism, completing a certain
number of executions, and restart policies. This KEP proposes an enhancement

> that will allow higher-level controllers to create Jobs with the option to NOT

Reviewer: ```suggestion
that will allow creating Jobs with the option to NOT
```

immediately create Pods until a bit is flipped. Such a Job is called a *Stopped
Job* and is facilitated through the `stopped` field in the Job spec.

## Motivation

The Job controller tracks and manages all Jobs within Kubernetes. When a new
Job is created and is seen by the Job controller, it will immediately begin
creating pods to satisfy the requirements of the Job.

> This is sometimes undesirable if a higher-level controller wants to create a

Reviewer: I think saying "higher-level controller wants to create" is confusing, the job will be managed by a higher-level controller, but creating the job could be done by users directly.

#### Story 2

I'm the system administrator of a cluster that can only run one Job at any
given time. I'd like my users to submit their Jobs whenever they want, but I
want control over whose Job gets to run on the cluster. With a custom mutating
webhook that adds `stopped: true` to all user-submitted Jobs, I can
indefinitely delay Pod creation.

I can write a higher-level Job queueing controller to unpause the right Job
based on external factors. For example, the controller could choose to simply
unpause Jobs in the FIFO order. Alternatively, Jobs could be assigned priority
and, just like kube-scheduler, the controller can make a decision based on the
stopped Job queue (including Job preemption). Each Job could request a
different amount of resources, so the higher-level controller may also want to
resize the cluster to just fit the Job it's going to run. Regardless of what
logic the controller uses to queue jobs, being able to stop Jobs indefinitely
and then resume them later is important. Note that this hypothetical controller
isn't necessary to do Job queueing; a human administrator could also make these

> decisions manually at a small enough scale.

Reviewer: I don't see that manual control of the jobs happening :)

#### Story 2

Stopping a running Job allows higher-level controllers to preempt Jobs. While
such a controller does not exist in-tree currently, Job priority and
preemption, when implemented and handled, could improve overall completion

> progress through better Job queueing. One can think of the Job preemption to be

Reviewer 4: Job queueing is more relevant in Story 1 than here.

there aren't enough resources or quota limits left in the cluster to
accommodate it, one or more low-priority Jobs can be stopped to free up
resources. After the cluster scales up or after the quota limits are raised,
the stopped Job can be resumed. The user can also make modifications to the
Job's mutable fields (e.g. parallelism) in the stopped state before resuming
it.

Note that stopping a Job for preemption is different from deleting and
re-creating it because stopping will retain the completion status of the Job.
The preempting controller can let a Job pick up from where it was stopped
without losing past completion data.

### Notes/Constraints/Caveats

While creating a stopped Job is well-understood, stopping an active Job
requires some careful thought. Assuming that the semantics of stopping a Job
are similar to that of pausing a Deployment, the controller manager must delete
all active Pods when a Job is stopped. Pods that successfully completed will
still count towards the Job's completions even if the Job is stopped.

Another interesting point of note is that it is important that system
administrators restricts users from accidentally modifying this field manually

> in a Job managed by a higher-level controller. This can be done with a mutating

Reviewer 4: validating or mutating webhook.

Also add webhooks to non-goals

Reviewer: what do you mean webhooks is not a goal?

> 1. Stopping an active Job will delete *all* active pods.

Reviewer 4: what are the mitigations?

I believe we can only offer documentation and best practices for both points.

Reviewer: why is this a risk, isn't this something we want by design?

Reviewer 4: It might be a risk depending on how the Pods are stopped and the completion semantics.
It needs to be clarified in docs.

Reviewer: sure, it needs to be clarified, but risks are unintended side effects, "delete all active pods" is not an unintended side effect, it is a design choice. So the risk here could be "unknown effects from deleting active pods when stopping a job"

Reviewer: unless of course we know the effects or impact on semantics, in which case we should mention them.

	// a Job is created in the stopped state, no Pods are created by the job
	// controller. If a Job enters the stopped state (i.e. the flag goes from
	// false to true), the job controller will delete all Pods associated with

>         // this job. This field is generally used by higher-level controllers that

Reviewer 4: nit: format looks off.

Specify the default value: false

But, I think this line shouldn't be part of the API documentation.
